Regression: error TS2322: Type OBJECT_LITERAL is not assignable to type 'Record<string, ...>'.

## Steps to reproduce

```js
export function test() {
	/** @type {Record<string, boolean>} */
	let merged_props = {};
  merged_props.foo = true;
  return merged_props;
}
```
[TSGO Playground](https://tsgo.sxzz.dev/#eNp1U8tu2zAQ/BWCp9YQ5Lv7QA/JIQGKAGlvllFQ1MphQ+0K5MpxYOjfsxQRw4pi6EBpZnY5+9BJt3qz3erOOCw56kK7rqfAqtLlmiFy+T9W+luFeldsNUdL2Lq9gISiPVWoRGmp652H8NCzIxT9RglTsTy6o2bwkKBKQ0Q4Cla8k97VidleUCJrqDu/lI4hmDpl2J3D2IQ9yPlpTqTbzk0khwHOcPbxCJH8kFzm6HrARoxfhIekOMC9FPj7bH2WKV/wgLchkEQu2YT/DQNa837RTGG8p5c/r8hPwM7eQGsGz3dT16fWzcQHCLWk6bKXFGWOC1Hk4Oyy5Omi+2VO+wT2eY7LGNMgrzQ2k1cmmcmFhQxDzMbvUMZI/YKngW/c1ESZtouXGRuw3oSPLczc1SElwVjhmLc1r6/sKRynnW5lJimfSsyXr3lJ16uV+sWvPajTI1gKzfdUC+4LVRN5MPhzVKt1Unpg1YF0qPnXSzFR/VCnUX4NNUPLlkiY5GbiAvAQcCYRXBzuCm31Rou9gxzeJFPy8aI3rfERxjePIj4/)

Today we got a bunch of such errors in the Svelte repo.

## Behavior with `typescript@6.0`

[No errors](https://www.typescriptlang.org/play/?jsx=0&ts=6.0.3&filetype=js#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVxsAZxgAoBKOAbwCgBIAegCoG4ABGATzGEoCVg1oAEwA8RKJhQBzADRwARhAgAbYAEMUAPgC+cBnVor4AW2BRJwQQH0wUCGAJwAvJS0BuKnDgmzF67fsAdAiKTnhQSMDunlDAMEhQuN7mVjZ2BO5aQA)

## Behavior with `tsgo`

Emits TS2322 error.

Using the unofficial [tsgo playground](https://tsgo.sxzz.dev) I've narrowed down the range to:
- As late as commit 2f6504c1b0efb1750abf948fe277981b50214093: No error matching 6.0
- At least by commit 6a6cb096b6b58e9c85ac7128def9fbe6cd5c42ab: `main.js(3,6): error TS2322: Type '{ foo: true; }' is not assignable to type 'Record<string, boolean>'. Index signature for type 'string' is missing in type '{ foo: true; }'.`

Which differs only by PR #3699 as per comparison https://github.com/microsoft/typescript-go/compare/2f6504c1b0efb1750abf948fe277981b50214093...6a6cb096b6b58e9c85ac7128def9fbe6cd5c42ab

--- a/internal/ast/utilities.go
+++ b/internal/ast/utilities.go
@@ -4077,7 +4077,7 @@ func HasModifier(node *Node, flags ModifierFlags) bool {
 	return node.ModifierFlags()&flags != 0
 }
 
-func IsExpandoInitializer(initializer *Node) bool {
+func IsExpandoInitializer(declaration *Node, initializer *Node) bool {
 	if initializer == nil {
 		return false
 	}
@@ -4085,7 +4085,7 @@ func IsExpandoInitializer(initializer *Node) bool {
 		return true
 	}
 	if IsInJSFile(initializer) {
-		return IsClassExpression(initializer) || (IsObjectLiteralExpression(initializer) && len(initializer.Properties()) == 0)
+		return IsClassExpression(initializer) || (IsObjectLiteralExpression(initializer) && len(initializer.Properties()) == 0 && declaration.Type() == nil)
 	}
 	return false
 }

--- a/internal/binder/binder.go
+++ b/internal/binder/binder.go
@@ -1108,20 +1108,20 @@ func getInitializerSymbol(symbol *ast.Symbol) *ast.Symbol {
 	declaration := symbol.ValueDeclaration
 	// For an assignment 'fn.xxx = ...', where 'fn' is a previously declared function or a previously
 	// declared const variable initialized with a function expression or arrow function, we add expando
-	// property declarations to the function's symbol.
-	// This also applies to class expressions and empty object literals in JS files.
+	// property declarations to the function's symbol. This also applies to class expressions in JS files,
+	// and empty object literals in JS files when the declaration doesn't have a type annotation.
 	switch {
 	case ast.IsFunctionDeclaration(declaration) || ast.IsInJSFile(declaration) && ast.IsClassDeclaration(declaration):
 		return symbol
 	case ast.IsVariableDeclaration(declaration) &&
 		(declaration.Parent.Flags&ast.NodeFlagsConst != 0 || ast.IsInJSFile(declaration)):
 		initializer := declaration.Initializer()
-		if ast.IsExpandoInitializer(initializer) {
+		if ast.IsExpandoInitializer(declaration, initializer) {
 			return initializer.Symbol()
 		}
 	case ast.IsBinaryExpression(declaration) && ast.IsInJSFile(declaration):
 		initializer := declaration.AsBinaryExpression().Right
-		if ast.IsExpandoInitializer(initializer) {
+		if ast.IsExpandoInitializer(declaration, initializer) {
 			return initializer.Symbol()
 		}
 	}

--- a/testdata/baselines/reference/compiler/nonExpandoDeclarations.symbols
+++ b/testdata/baselines/reference/compiler/nonExpandoDeclarations.symbols
@@ -1,0 +1,36 @@
+//// [tests/cases/compiler/nonExpandoDeclarations.ts] ////
+
+=== main.js ===
+function f1() {
+>f1 : Symbol(f1, Decl(main.js, 0, 0))
+
+    /** @type {Record<string, boolean>} */
+    let merged_props = {};
+>merged_props : Symbol(merged_props, Decl(main.js, 2, 7))
+
+    merged_props.foo = true;
+>merged_props : Symbol(merged_props, Decl(main.js, 2, 7))
+
+    return merged_props;
+>merged_props : Symbol(merged_props, Decl(main.js, 2, 7))
+}
+
+function f2() {}
+>f2 : Symbol(f2, Decl(main.js, 5, 1))
+
+/** @type {Record<string, boolean>} */
+f2.merged_props = {};
+>f2.merged_props : Symbol(f2.merged_props, Decl(main.js, 7, 16))
+>f2 : Symbol(f2, Decl(main.js, 5, 1))
+>merged_props : Symbol(f2.merged_props, Decl(main.js, 7, 16))
+
+f2.merged_props.foo = true;
+>f2.merged_props : Symbol(f2.merged_props, Decl(main.js, 7, 16))
+>f2 : Symbol(f2, Decl(main.js, 5, 1))
+>merged_props : Symbol(f2.merged_props, Decl(main.js, 7, 16))
+
+f2.merged_props
+>f2.merged_props : Symbol(f2.merged_props, Decl(main.js, 7, 16))
+>f2 : Symbol(f2, Decl(main.js, 5, 1))
+>merged_props : Symbol(f2.merged_props, Decl(main.js, 7, 16))
+

--- a/testdata/baselines/reference/compiler/nonExpandoDeclarations.types
+++ b/testdata/baselines/reference/compiler/nonExpandoDeclarations.types
@@ -1,0 +1,47 @@
+//// [tests/cases/compiler/nonExpandoDeclarations.ts] ////
+
+=== main.js ===
+function f1() {
+>f1 : () => Record<string, boolean>
+
+    /** @type {Record<string, boolean>} */
+    let merged_props = {};
+>merged_props : Record<string, boolean>
+>{} : {}
+
+    merged_props.foo = true;
+>merged_props.foo = true : true
+>merged_props.foo : boolean
+>merged_props : Record<string, boolean>
+>foo : boolean
+>true : true
+
+    return merged_props;
+>merged_props : Record<string, boolean>
+}
+
+function f2() {}
+>f2 : { (): void; merged_props: Record<string, boolean>; }
+
+/** @type {Record<string, boolean>} */
+f2.merged_props = {};
+>f2.merged_props = {} : {}
+>f2.merged_props : Record<string, boolean>
+>f2 : { (): void; merged_props: Record<string, boolean>; }
+>merged_props : Record<string, boolean>
+>{} : {}
+
+f2.merged_props.foo = true;
+>f2.merged_props.foo = true : true
+>f2.merged_props.foo : boolean
+>f2.merged_props : Record<string, boolean>
+>f2 : { (): void; merged_props: Record<string, boolean>; }
+>merged_props : Record<string, boolean>
+>foo : boolean
+>true : true
+
+f2.merged_props
+>f2.merged_props : Record<string, boolean>
+>f2 : { (): void; merged_props: Record<string, boolean>; }
+>merged_props : Record<string, boolean>
+

--- a/testdata/tests/cases/compiler/nonExpandoDeclarations.ts
+++ b/testdata/tests/cases/compiler/nonExpandoDeclarations.ts
@@ -1,0 +1,19 @@
+// @checkJs: true
+// @noEmit: true
+
+// https://github.com/microsoft/typescript-go/issues/3709
+
+// @filename: main.js
+
+function f1() {
+    /** @type {Record<string, boolean>} */
+    let merged_props = {};
+    merged_props.foo = true;
+    return merged_props;
+}
+
+function f2() {}
+/** @type {Record<string, boolean>} */
+f2.merged_props = {};
+f2.merged_props.foo = true;
+f2.merged_props

